package dolos;

message quaternion {
    required float e0 = 1;
    required float ex = 2;
    required float ey = 3;
    required float ez = 4;
}

message load_factor {
    required float nx = 1;
    required float ny = 2;
    required float nz = 3;
}

message angular_rates {
    required float p = 1;
    required float q = 2;
    required float r = 3;
}

message daq_measurement {
    required float n_x = 1;
    required float n_y = 2;
    required float n_z = 3;
    required float p = 4;
    required float q = 5;
    required float r = 6;
    required float v_inf = 7;
    required float h = 8;
}

message imu_vars {
    required load_factor load = 1;
    required angular_rates rates = 2;
    required quaternion quat = 3;
}

message atmosphere_vars {
    required float p = 1;
    required float T = 2;
}

message gps_vars {
    optional float lat = 1;
    optional float lon = 2;
    optional float alt = 3;
    optional float spd = 4;
    optional float climb = 5;
    optional float track = 6;
}

message pitot_vars {
    required float q = 1;
    optional float p = 2;
}

message sensor_measurement {
    optional imu_vars imu = 1;
    optional atmosphere_vars atmos = 2;
    optional gps_vars gps = 3;
    optional pitot_vars pitot = 4;
}

message state_vars {
    optional load_factor load = 1;
    optional quaternion quat = 2;
    optional angular_rates rates = 3;
    optional float v_inf = 4;
    optional float h = 5;
    optional float x = 6;
    optional float y = 7;
    optional float T = 8;
    optional float p = 9;
    optional float qnh = 10;
}

message actuation_vars {
    required float d_a = 1;
    required float d_e = 2;
    required float d_r = 3;
    required float motor_pwr = 4;
}

message aircraft_state {
    required state_vars ac_state = 1;
    required actuation_vars control_state = 2;
}

message setpoint {
    required float n_z = 1;
    required float n_x = 2;
    required float q = 3;
    required float motor_pwr = 4;
}

message location {
    required double lat = 1;
    required double lon = 2;
    required float alt = 3;
}

message heartbeat {
    required int32 epoch_low = 1;
    required int32 epoch_hi = 2;
    required int32 millis = 3;
}

message jam_ctl {
    required bool state = 1;
    optional float freq = 2;
}

message control_packet {
    required heartbeat hb = 1;
    optional setpoint input = 2;
    optional actuation_vars direct = 3;
    optional location autopilot_loc = 4;
    optional jam_ctl jammer_cmd = 5;
}

message telemetry_packet {
    required hearbeat hb = 1;
}

message actuation_packet {
    optional actuation_vars direct = 1;
    optional setpoint input = 2;
}
